### ISSUE TITLE
wm starts with inactive monitor focused

### ISSUE THREAD

**Issue by TLINDEN** [herbstluftwm/herbstluftwm#1363]
Howdy,

I have a notebook which is connected to a wide screen monitor. I am using the `hc detect_monitors` setting.

The lid of the notebook is closed when I boot, I can login as usual, then herbstluftwm starts and shows tag 0. However, the widescreen monitor is not in focus, it's the notebook screen.

So on startup I always have to switch to the wide screen once. Yes, this is a really minor annoyance of an otherwise excellent WM (I'm coming from xmonad and I LOVE herbstluftwm already!).

However, is there something I can configure so that the active monitor (if theres only one active) gets the focus on startup?

```
scip@tripod: % xrandr --query
Screen 0: minimum 320 x 200, current 5360 x 1440, maximum 16384 x 16384
eDP-1 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 344mm x 193mm
   1920x1080     60.00*+  59.97    59.96    59.93    40.00  
   1680x1050     59.95    59.88  
   1600x1024     60.17  
   1400x1050     59.98  
   1600x900      59.99    59.94    59.95    59.82  
   1280x1024     60.02  
   1440x900      59.89  
   1400x900      59.96    59.88  
   1280x960      60.00  
   1440x810      60.00    59.97  
   1368x768      59.88    59.85  
   1360x768      59.80    59.96  
   1280x800      59.99    59.97    59.81    59.91  
   1152x864      60.00  
   1280x720      60.00    59.99    59.86    59.74  
   1024x768      60.04    60.00  
   960x720       60.00  
   928x696       60.05  
   896x672       60.01  
   1024x576      59.95    59.96    59.90    59.82  
   960x600       59.93    60.00  
   960x540       59.96    59.99    59.63    59.82  
   800x600       60.00    60.32    56.25  
   840x525       60.01    59.88  
   864x486       59.92    59.57  
   800x512       60.17  
   700x525       59.98  
   800x450       59.95    59.82  
   640x512       60.02  
   720x450       59.89  
   700x450       59.96    59.88  
   640x480       60.00    59.94  
   720x405       59.51    58.99  
   684x384       59.88    59.85  
   680x384       59.80    59.96  
   640x400       59.88    59.98  
   576x432       60.06  
   640x360       59.86    59.83    59.84    59.32  
   512x384       60.00  
   512x288       60.00    59.92  
   480x270       59.63    59.82  
   400x300       60.32    56.34  
   432x243       59.92    59.57  
   320x240       60.05  
   360x202       59.51    59.13  
   320x180       59.84    59.32  
DP-1 disconnected (normal left inverted right x axis y axis)
HDMI-1 connected 3440x1440+1920+0 (normal left inverted right x axis y axis) 797mm x 333mm
   3440x1440     49.99*+
   2560x1440     59.95  
   2560x1080     60.00    59.94  
   1920x1080     60.00    50.00    59.94  
   1680x1050     59.88  
   1600x900      60.00  
   1280x1024     75.02    60.02  
   1440x900      59.90  
   1280x800      59.91  
   1152x864      75.00  
   1280x720      60.00    50.00    59.94  
   1024x768      75.03    70.07    60.00  
   832x624       74.55  
   800x600       72.19    75.00    60.32    56.25  
   720x576       50.00  
   720x480       60.00    59.94  
   640x480       75.00    72.81    66.67    60.00    59.94  
   720x400       70.08  
DP-2 disconnected (normal left inverted right x axis y axis)
HDMI-2 disconnected (normal left inverted right x axis y axis)
```

I suspect the behavior is being caused by the notebook screen set to *primary*?

Thanks in advance,
Tom

**Comment by t-wissmann** [comment]
By the way, according to the `xrandr` man page, the primary monitor is simply the monitor first in the list:
```
       --primary
              Set the output as primary.  It will be sorted first in Xinerama and RANDR geometry requests.
```
So you are entirely right that the issue is that your notebook screen is set to primary. Maybe you can find a way (e.g. a script calling `xrandr …`) such that the external monitor is the primary screen (this should happen before `detect_monitors` is called, so e.g. in the `.xinitrc` or `.xsession`, depending on what your precise setup is).

However, I can't help much from the hlwm side.

**Comment by TLINDEN** [comment]
Hi,

this doesn't work here. I have set the monitor on HDMI1 to primary via `.xsessionrc`, it is also being displayed as primary:
```
% xrandr --query | egrep "^[A-Za-z].* connected"
eDP-1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 344mm x 193mm
HDMI-1 connected primary 3440x1440+1920+0 (normal left inverted right x axis y axis) 797mm x 333mm
```

So, the HDMI monitor is primary but still not the first in the list and the behavior is as already reported. Btw, I also tried to just disable the notebook monitor, but then the screen on HDMI1 looks very strangs. And it's dangerous in case something happens to the monitor, in such a case I couldn't use the notebook anymore.

So, I'll keep it as it is and live with it.

Thank you for your help and effort anyway!

**Comment by t-wissmann** [comment]
TL;DR: there is a hlwm bug: hlwm sorts the monitors by their geometry in order to remove duplicates. I will fix it soon.

Long version:
The order of the `xrandr --query` output is not how it is reported to the window manager. The order of outputs is reported by `--listmonitors`:
```
$ xrandr --output DP-2 --primary --mode 1920x1200 --pos 0x0 --output HDMI-0 --mode 1280x1024 --pos 1920x0                           
$ xrandr --query | egrep "^[A-Za-z].* connected"                                                                                    
DP-2 connected primary 1920x1200+0+0 (normal left inverted right x axis y axis) 518mm x 324mm
HDMI-0 connected 1280x1024+1920+0 (normal left inverted right x axis y axis) 376mm x 301mm
$ xrandr --listmonitors                                                                                                             
Monitors: 2
 0: +*DP-2 1920/518x1200/324+0+0  DP-2
 1: +HDMI-0 1280/376x1024/301+1920+0  HDMI-0
$ hc detect_monitors -l                                                                                                             
1920x1200+0+0
1280x1024+1920+0
$ # now, let us make the HDMI output primary                                                                                        
$ xrandr --output DP-2 --mode 1920x1200 --pos 0x0 --output HDMI-0 --mode 1280x1024 --pos 1920x0 --primary                           
$ xrandr --query | egrep "^[A-Za-z].* connected" # here, the order does not change                                                  
DP-2 connected 1920x1200+0+0 (normal left inverted right x axis y axis) 518mm x 324mm
HDMI-0 connected primary 1280x1024+1920+0 (normal left inverted right x axis y axis) 376mm x 301mm
$ xrandr --listmonitors                                                                                                             
Monitors: 2
 0: +*HDMI-0 1280/376x1024/301+1920+0  HDMI-0
 1: +DP-2 1920/518x1200/324+0+0  DP-2
$                                                                                                                                   
```
Note how the monitor with the offset `+1920+0` is now listed first. Unfortunately, this order is destroyed by `detect_monitors`:
```
$ hc detect_monitors -l                                                                                                             
1920x1200+0+0
1280x1024+1920+0
```
This is simply a bug, which should be fixed soon. Sorry for the trouble!

**Comment by t-wissmann** [comment]
This issue is now solved in PR #1427 (and soon in master). Sorry again for not considering that this was a hlwm bug.

**Comment by TLINDEN** [comment]
Ha, amazing :) Thanks for reconsidering the issue!
